drop-transfer/src/ws/server/mod.rs

```diff
@@ -310,51 +310,58 @@ async fn handle_client(
     let _guard = TransferGuard::new(state.clone(), xfer.id());
     let (req_send, mut req_rx) = mpsc::unbounded_channel();
 
-    {
-        if let Err(err) = state
+    let task = async {
+        state
             .transfer_manager
             .lock()
             .await
             .insert_transfer(xfer.clone(), TransferConnection::Server(req_send.clone()))
-        {
-            error!(logger, "Failed to insert a new trasfer: {}", err);
-
-            let _ = handler
-                .on_error(
-                    &mut socket,
-                    anyhow::anyhow!("Failed to init transfer: {err}"),
-                )
-                .await;
-            let _ = socket.close().await;
-
-            return;
-        }
-    }
-
-    let mode = match state.storage.insert_transfer(&xfer.storage_info()).await {
-        Ok(mode) => mode,
-        Err(err) => {
-            error!(logger, "Failed to insert transfer into storage: {err}",);
-            drop_storage::TransferIncomingMode::New
-        }
-    };
+            .context("Failed to insert a new transfer")?;
 
-    match mode {
-        drop_storage::TransferIncomingMode::New => {
-            state
-                .event_tx
-                .send(Event::RequestReceived(xfer.clone()))
-                .await
-                .expect("Failed to notify receiving peer!");
-        }
-        drop_storage::TransferIncomingMode::Resume => {
-            let task = async {
-                info!(
+        let existing_info = match state.storage.incoming_transfer_info(xfer.id()).await {
+            Ok(info) => info,
+            Err(err) => {
+                error!(
                     logger,
-                    "Transfer {} resume. Resuming started files",
-                    xfer.id()
+                    "DB failed upon checking if transfer is a resume: {err:?}",
                 );
+                // Fall back to treating it as a new one
+                None
+            }
+        };
+
+        let current_info = xfer.storage_info();
+        if let Some(mut existing_info) = existing_info {
+            // Check if the transfer matches
+            anyhow::ensure!(
+                current_info.peer == existing_info.peer,
+                "Transfer {} resume. Peers do not match",
+                xfer.id()
+            );
```

Comment:
This whole close is very difficult to read and validate that it is correct, could you split it into multiple, smaller parts?

Reply:
Yes, let me try

Reply:
Refactored a bit. Can you take a look?